orbit/pkg/luks/luks_linux.go

```diff
@@ -0,0 +1,264 @@
+
+func (lr *LuksRunner) Run(oc *fleet.OrbitConfig) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	if !oc.Notifications.RunDiskEncryptionEscrow {
+		return nil
+	}
+
+	devicePath, err := lvm.FindRootDisk()
+	if err != nil {
+		return fmt.Errorf("Failed to find LUKS Root Partition: %w", err)
+	}
+
+	var response LuksResponse
+	key, keyslot, err := lr.getEscrowKey(ctx, devicePath)
+	if err != nil {
+		response.Err = err.Error()
+	}
+
+	response.Passphrase = string(key)
+	response.KeySlot = keyslot
+
+	if keyslot != nil {
+		salt, err := getSaltforKeySlot(ctx, devicePath, *keyslot)
+		if err != nil {
+			return fmt.Errorf("Failed to get salt for key slot: %w", err)
+		}
+		response.Salt = salt
+	}
+
+	if err := lr.escrower.SendLinuxKeyEscrowResponse(response); err != nil {
+		if err := lr.infoPrompt(ctx, infoFailedTitle, infoFailedText); err != nil {
+			log.Debug().Err(err).Msg("failed to show failed escrow key dialog")
+		}
+		return fmt.Errorf("escrower escrowKey err: %w", err)
+	}
+
+	if response.Err != "" {
+		if err := lr.infoPrompt(ctx, infoFailedTitle, response.Err); err != nil {
+			log.Debug().Err(err).Msg("failed to show failed escrow key dialog")
+		}
+		return fmt.Errorf("error getting linux escrow key: %s", response.Err)
+	}
+
+	// Show success dialog
+	if err := lr.infoPrompt(ctx, infoSuccessTitle, infoSuccessText); err != nil {
+		log.Debug().Err(err).Msg("failed to show success escrow key dialog")
+	}
+
+	return nil
+}
+
+func (lr *LuksRunner) getEscrowKey(ctx context.Context, devicePath string) ([]byte, *uint, error) {
+	device := luksdevice.New(luksdevice.AESXTSPlain64Cipher)
+
+	// Prompt user for existing LUKS passphrase
+	passphrase, err := lr.entryPrompt(ctx, entryDialogTitle, entryDialogText)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to show passphrase entry prompt: %w", err)
+	}
+
+	// Validate the passphrase
+	for {
+		valid, err := lr.passphraseIsValid(ctx, device, devicePath, passphrase)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Failed validating passphrase: %w", err)
+		}
+
+		if !valid {
+			passphrase, err = lr.entryPrompt(ctx, entryDialogTitle, retryEntryDialogText)
+			if err != nil {
+				return nil, nil, fmt.Errorf("Failed re-prompting for passphrase: %w", err)
+			}
+			continue
+		}
+
+		break
+	}
+
+	if len(passphrase) == 0 {
+		log.Debug().Msg("Passphrase is empty, no password supplied, dialog was canceled, or timed out")
+		return nil, nil, nil
+	}
+
+	escrowPassphrase, err := generateRandomPassphrase()
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to generate random passphrase: %w", err)
+	}
+
+	// Create a new key slot, error if all key slots are full
+	// keySlot 0 is assumed to be the user's passphrase
+	// so we start at 1
+	var keySlot uint = 1
+	for {
+		if keySlot == maxKeySlots {
+			return nil, nil, errors.New("all LUKS key slots are full")
+		}
+
+		userKey := encryption.NewKey(0, passphrase)
```

Comment:
We should document that if the user has two keyslots, the the passphrase we are prompting is the `0` one. (Not sure how often do users have multiple..., so just documenting is probably enough for now)